receiver/integrationreceiver/receiver.go

```diff
@@ -0,0 +1,291 @@
+		if consumerChain.traces != nil {
+			traces, err := factory.CreateTraces(ctx, params, config, consumerChain.traces)
+			if err != nil {
+				return fmt.Errorf("failed to create traces processor %s: %w", params.ID, err)
+			}
+			consumerChain.traces = traces
+			components = append(components, traces)
+		}
+	}
+
+	params := r.params
+	params.ID = component.NewIDWithName(receiverFactory.Type(), fmt.Sprintf("%s-receiver", pipelineID))
+	params.Logger = params.Logger.With(zap.String("name", params.ID.String()))
+	receiversCreated := 0
+	if consumerChain.logs != nil {
+		logs, err := receiverFactory.CreateLogs(ctx, params, preparedConfig, consumerChain.logs)
+		switch {
+		case err == nil:
+			components = append(components, logs)
+			receiversCreated += 1
+		case errors.Is(err, otelpipeline.ErrSignalNotSupported):
+			r.params.Logger.Debug("receiver does not support logs telemetry type",
+				zap.String("integration", r.params.ID.String()),
+				zap.String("receiver", params.ID.String()))
+		default:
+			return fmt.Errorf("failed to create logs receiver %s: %w", params.ID, err)
+		}
+	}
+	if consumerChain.metrics != nil {
+		metrics, err := receiverFactory.CreateMetrics(ctx, params, preparedConfig, consumerChain.metrics)
+		switch {
+		case err == nil:
+			components = append(components, metrics)
+			receiversCreated += 1
+		case errors.Is(err, otelpipeline.ErrSignalNotSupported):
+			r.params.Logger.Debug("receiver does not support metrics telemetry type",
+				zap.String("integration", r.params.ID.String()),
+				zap.String("receiver", params.ID.String()))
+		default:
+			return fmt.Errorf("failed to create metrics receiver %s: %w", params.ID, err)
+		}
+	}
+	if consumerChain.traces != nil {
+		traces, err := receiverFactory.CreateTraces(ctx, params, preparedConfig, consumerChain.traces)
+		switch {
+		case err == nil:
+			components = append(components, traces)
+			receiversCreated += 1
+		case errors.Is(err, otelpipeline.ErrSignalNotSupported):
+			r.params.Logger.Debug("receiver does not support traces telemetry type",
+				zap.String("integration", r.params.ID.String()),
+				zap.String("receiver", params.ID.String()))
+		default:
+			return fmt.Errorf("failed to create traces receiver %s: %w", params.ID, err)
+		}
+	}
+
+	// If no receiver has been created the rest of the pipeline won't be used, so don't keep it.
+	if receiversCreated == 0 {
+		// Shutting down created components out of kindness, because they haven't been started yet.
+		if err := shutdownComponents(ctx, components); err != nil {
+			r.params.Logger.Error("failed to cleanup processors after receiver was not created",
+				zap.String("integration", r.params.ID.String()),
+				zap.String("receiver", params.ID.String()))
+		}
+		return nil
+	}
+
+	for _, component := range components {
+		err := component.Start(ctx, host)
+		if err != nil {
+			return fmt.Errorf("failed to start component %q: %w", component, err)
+		}
+	}
+	r.components = append(r.components, components...)
```

Comment:
Should we append the components to the receiver's state as soon as are being started? There might be some components that won't be shutdown if there's an error after the first started component.

Reply:
Sounds like a good idea, yes, I will do it.